core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala

```diff
@@ -892,27 +911,30 @@ private[spark] class PythonBroadcast(@transient var path: String) extends Serial
   private def readObject(in: ObjectInputStream): Unit = Utils.tryOrIOException {
     val dir = new File(Utils.getLocalDir(SparkEnv.get.conf))
     val file = File.createTempFile("broadcast", "", dir)
+    phantomReferences += new FilePhantomReference(file, queue)
     path = file.getAbsolutePath
     val out = new FileOutputStream(file)
     Utils.tryWithSafeFinally {
       Utils.copyStream(in, out)
     } {
-      out.close()
+       in.close();
     }
   }
 
-  /**
-   * Delete the file once the object is GCed.
+  /** Create a seperate daemon thread
+   *  to remove phantomreferences from queue and invoke cleanup
    */
-  override def finalize() {
-    if (!path.isEmpty) {
-      val file = new File(path)
-      if (file.exists()) {
-        if (!file.delete()) {
-          logWarning(s"Error deleting ${file.getPath}")
-        }
+    val referenceThread = new Thread(threadName){
```

Comment:
This will create a thread for each `PythonBroadcast`. We should avoid that.

Maybe use a global thread to do that? However, this will leak a thread even if it's not used any more (e.g., after stopping SparkContext).

Maybe `finalize` is the best option.